crates/router/src/core/payments.rs

```diff
@@ -2546,6 +2548,76 @@ pub async fn get_filters_for_payments(
     ))
 }
 
+#[cfg(feature = "olap")]
+pub async fn get_payment_filters(
+    state: AppState,
+    merchant: domain::MerchantAccount,
+) -> RouterResponse<api::PaymentListFiltersV2> {
+    use std::collections::{HashMap, HashSet};
+
+    use api_models::admin::MerchantConnectorInfo;
+
+    use super::admin::list_payment_connectors;
+    use crate::services::ApplicationResponse;
+
+    let merchant_connector_accounts = if let ApplicationResponse::Json(data) =
+        list_payment_connectors(state, merchant.merchant_id).await?
+    {
+        data
+    } else {
+        return Err(errors::ApiErrorResponse::InternalServerError.into());
+    };
+
+    let mut connector_map: HashMap<String, Vec<MerchantConnectorInfo>> = HashMap::new();
+    let mut payment_method_types_map: HashMap<
+        enums::PaymentMethod,
+        HashSet<enums::PaymentMethodType>,
+    > = HashMap::new();
+
+    for merchant_connector_account in merchant_connector_accounts {
+        if let Some(label) = merchant_connector_account.connector_label {
+            let info = MerchantConnectorInfo {
+                connector_label: label,
+                merchant_connector_id: merchant_connector_account.merchant_connector_id,
+            };
+
+            connector_map
+                .entry(merchant_connector_account.connector_name.clone())
+                .or_insert_with(Vec::new)
+                .push(info);
+        }
+
+        // Populate all the enabled payment methods and corresponding payment method types
+        if let Some(ref payment_methods_enabled) =
+            merchant_connector_account.payment_methods_enabled
+        {
+            for payment_method_enabled in payment_methods_enabled {
+                if let Some(payment_method_types_vec) = &payment_method_enabled.payment_method_types
+                {
+                    let payment_method = payment_method_enabled.payment_method;
+                    let types_set = payment_method_types_map
+                        .entry(payment_method.clone())
+                        .or_insert_with(HashSet::new);
+
+                    for request_payment_method_type in payment_method_types_vec {
+                        types_set.insert(request_payment_method_type.payment_method_type.clone());
+                    }
+                }
+            }
+        }
+    }
```

Comment:
This could be done better by extracting connectors and payment methods separately. And use iterators to reduce the nesting.